### PR TITLE
Add a conditions lint rule + other rules fixes.

### PR DESCRIPTION
**Pull request by pchote**
The rule changes are mainly reorganizing inherits to fix cases where conditions are granted but not used or visa versa.  This also sneaks in the TS crate stat bonus changes @GraionDilach requested in #12381, and supersedes #12476.

### CODE REVIEW

File: OpenRA.Mods.Common/Lint/LintExts.cs [hunk 1]
```diff
@@ -24,20 +25,41 @@ public static IEnumerable<string> GetFieldValues(object ruleInfo, FieldInfo fiel
...
+			if (typeof(IEnumerable).IsAssignableFrom(type))
+				return (IEnumerable<string>)propertyInfo.GetValue(ruleInfo);
+
+			if (type == typeof(BooleanExpression))
+			{
+				var expr = (BooleanExpression)propertyInfo.GetValue(ruleInfo);
+				return expr != null ? expr.Variables : Enumerable.Empty<string>();
+			}
+
+			emitError("Bad type for reference on {0}.{1}. Supported types: string, string[], HashSet<string>, BooleanExpression"
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Lint/LintExts.cs]: Is there a reason this message wasn't changed?

File: OpenRA.Game/Traits/LintAttributes.cs [hunk 2]
```diff
@@ -46,11 +46,11 @@ public SequenceReferenceAttribute(string imageReference = null, bool prefix = fa
...
 
-	[AttributeUsage(AttributeTargets.Field)]
-	public sealed class UpgradeGrantedReferenceAttribute : Attribute { }
+	[AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
+	public sealed class GrantedConditionReferenceAttribute : Attribute { }
 
-	[AttributeUsage(AttributeTargets.Field)]
-	public sealed class UpgradeUsedReferenceAttribute : Attribute { }
+	[AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
+	public sealed class ConsumedConditionReference : Attribute { }
```
Comment (reaperrr) [inline OpenRA.Game/Traits/LintAttributes.cs]: Is this intentional? Just `*Reference` here but `*ReferenceAttribute` above?

File: mods/ts/rules/gdi-vehicles.yaml [hunk 3]
```diff
@@ -1,5 +1,7 @@
 APC:
-	Inherits: ^VoxelTank
+	Inherits: ^Tank
+	Inherits@VOXELS: ^VoxelActor
+	Inherits@EXPERIENCE: ^GainsExperience
```
Comment (reaperrr) [inline mods/ts/rules/gdi-vehicles.yaml]: The TS APC is unarmed.

File: mods/ts/rules/nod-vehicles.yaml [hunk 4]
```diff
@@ -191,7 +199,9 @@ ART2:
 		Voice: Move
 
 REPAIR:
-	Inherits: ^VoxelTank
+	Inherits: ^Tank
+	Inherits@VOXELS: ^VoxelActor
+	Inherits@EXPERIENCE: ^GainsExperience
```
Comment (reaperrr) [inline mods/ts/rules/nod-vehicles.yaml]: The Repair vehicle shouldn't gain experience.

File: mods/ts/rules/nod-vehicles.yaml [hunk 5]
```diff
@@ -255,13 +266,14 @@ WEED:
...
-	-GainsExperience:
 	-DamagedByTerrain@VEINS:
 	-GrantConditionOnTerrain@VEINS:
 	-WithIdleOverlay@VEINS:
 
 SAPC:
-	Inherits: ^VoxelTank
+	Inherits: ^Tank
+	Inherits@VOXELS: ^VoxelActor
+	Inherits@EXPERIENCE: ^GainsExperience
```
Comment (reaperrr) [inline mods/ts/rules/nod-vehicles.yaml]: The subterranean APC is unarmed.

**Comment by pchote** [comment]
Fixed.

**Comment by GraionDilach** [comment]
Eh, unarmed crushers can still gain experience by crushing - although I haven't looked at the TS inis to confirm if they allow such.

**Comment by reaperrr** [comment]
> Eh, unarmed crushers can still gain experience by crushing

Well, they did in the original, but can vehicles in OpenRA gain experience from crushing?

**Comment by GraionDilach** [comment]
Sure they can, https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/Crushable.cs#L61 passes the crusher along.

**Comment by pchote** [comment]
So, what do we want to do here, then?

**Comment by GraionDilach** [comment]
On a second check, the above discussion has no change on the previous requests because no affected vehicle can actually crush any infantry.

**Review by GraionDilach — APPROVED** [review]

**Comment by GraionDilach** [comment]
What's left besides this, btw? Besides the obvious UpgradeManager->ConditionManager and the Upgrades folder in the code structure renames...

**Comment by pchote** [comment]
Fixing the internal naming should now be the last step.

**Comment by GraionDilach** [comment]
:+1: anyway, but I forgot to indicate it in the actual review - as I tend to do once in every 5 cases. >_>

File: OpenRA.Mods.Common/Lint/LintExts.cs [hunk 6]
```diff
@@ -24,20 +25,35 @@ public static IEnumerable<string> GetFieldValues(object ruleInfo, FieldInfo fiel
...
+				return (IEnumerable<string>)propertyInfo.GetValue(ruleInfo);
+
+			if (type == typeof(BooleanExpression))
+			{
+				var expr = (BooleanExpression)propertyInfo.GetValue(ruleInfo);
+				return expr != null ? expr.Variables : Enumerable.Empty<string>();
+			}
 
-			return new string[] { };
+			throw new InvalidOperationException("Bad type for reference on {0}.{1}. Supported types: string, IEnumerable<string>, BooleanExpression");
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Lint/LintExts.cs]: Am I missing some here? There are no parameters for those two substitutions.

**Comment by pchote** [comment]
Fixed and rebased.

**Review by obrakmann — APPROVED** [review]
👍

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d347b671eccdeae442d9c89191df8e9280efa403...7c0786fcc2d8ef5b63b8ccd87e3a6b5ba68bcab1)

**Pull request merged**